storage/database/rocksdb_database.go

```diff
@@ -0,0 +1,399 @@
+		}
+	} else if os.IsNotExist(err) && needToMake {
+		if err := os.MkdirAll(path, 0o755); err != nil {
+			return err
+		}
+	} else {
+		return err
+	}
+
+	return nil
+}
+
+func NewRocksDB(path string, config *RocksDBConfig) (*rocksDB, error) {
+	localLogger := logger.NewWith("path", path)
+
+	if err := openFile(path, !config.Secondary); err != nil {
+		return nil, err
+	}
+
+	blockCacheSize := config.CacheSize / 2 * 1024 * 1024 // half of cacheSize in MiB
+	bufferSize := config.CacheSize / 2 * 1024 * 1024     // half of cacheSize in MiB
+
+	bbto := grocksdb.NewDefaultBlockBasedTableOptions()
+	bbto.SetBlockCache(grocksdb.NewLRUCache(blockCacheSize))
+	policy := filterPolicyStrToNative(config.FilterPolicy)
+	if policy != nil {
+		bbto.SetFilterPolicy(policy)
+		bbto.SetOptimizeFiltersForMemory(true)
+	}
+
+	opts := grocksdb.NewDefaultOptions()
+	opts.SetBlockBasedTableFactory(bbto)
+	opts.SetCreateIfMissing(true)
+	opts.SetWriteBufferSize(bufferSize)
+	opts.SetDumpMallocStats(config.DumpMallocStat)
+	opts.SetCompression(compressionStrToType(config.CompressionType))
+	opts.SetBottommostCompression(compressionStrToType(config.BottommostCompressionType))
+
+	logger.Info("RocksDB configuration", "blockCacheSize", blockCacheSize, "bufferSize", bufferSize, "enableDumpMallocStat", config.DumpMallocStat, "compressionType", config.CompressionType, "bottommostCompressionType", config.BottommostCompressionType, "filterPolicy", config.FilterPolicy, "disableMetrics", config.DisableMetrics)
+
+	var (
+		db  *grocksdb.DB
+		err error
+	)
+
+	if config.Secondary {
+		db, err = grocksdb.OpenDbAsSecondary(opts, path, path)
+	} else {
+		db, err = grocksdb.OpenDb(opts, path)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return &rocksDB{
+		config: config,
+		db:     db,
+		wo:     grocksdb.NewDefaultWriteOptions(),
+		ro:     grocksdb.NewDefaultReadOptions(),
+		logger: localLogger,
+		quitCh: make(chan struct{}),
+	}, nil
+}
+
+func (db *rocksDB) Type() DBType {
+	return RocksDB
+}
+
+func (db *rocksDB) Put(key []byte, value []byte) error {
+	if db.config.Secondary {
+		return nil
+	}
+	if !db.config.DisableMetrics {
+		start := time.Now()
+		defer db.putTimer.Update(time.Since(start))
+	}
+	return db.db.Put(db.wo, key, value)
+}
+
+func (db *rocksDB) Has(key []byte) (bool, error) {
+	if db.config.Secondary {
+		db.db.TryCatchUpWithPrimary()
+	}
+
+	dat, err := db.db.GetBytes(db.ro, key)
+	if dat == nil || err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (db *rocksDB) Get(key []byte) ([]byte, error) {
+	if db.config.Secondary {
+		db.db.TryCatchUpWithPrimary()
+	}
+	if !db.config.DisableMetrics {
+		start := time.Now()
+		defer db.getTimer.Update(time.Since(start))
+	}
+	return db.get(key)
+}
+
+func (db *rocksDB) get(key []byte) ([]byte, error) {
+	dat, err := db.db.GetBytes(db.ro, key)
+	if dat == nil {
+		return nil, dataNotFoundErr
+	}
+
+	if err != nil {
+		return nil, err
+	}
+	return dat, nil
+}
```

Comment:
Don't we need to call `db.db.TryCatchUpWithPrimary()` for this? The method isn't used anywhere, though.

Reply:
There is `Get` method. `get` method is in there. It won't be searched if you use IDE because of build tag I guess.